packages/fxa-settings/src/lib/integrations/integration-factory-flags.ts

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Constants } from '../constants';
+import { ModelDataStore, UrlQueryData } from '../model-data';
+import { IntegrationFlags } from './interfaces/integration-flags';
+
+const DEVICE_PAIRING_SUPPLICANT_PATHNAME_REGEXP = /^\/pair\/supp/;
+
+/**
+ * Extrapolates flags from the state of the current data store. The collective state of these flags are used by
+ * the factory to determine what underlying type of integration to create.
+ *
+ * Note: this logic was ported from fxa-content-server app-start.js.
+ */
+export class DefaultIntegrationFactoryFlags implements IntegrationFlags {
+  protected get pathname() {
+    return this.urlQueryData.pathName;
+  }
+  constructor(private urlQueryData: UrlQueryData) {}
+
+  // TODO: Same method on the Relier flags, can we DRY?
+  isDevicePairingAsAuthority() {
+    return (
+      this._searchParam('redirect_uri') ===
+      Constants.DEVICE_PAIRING_AUTHORITY_REDIRECT_URI
+    );
+  }
+
+  isDevicePairingAsSupplicant() {
+    return DEVICE_PAIRING_SUPPLICANT_PATHNAME_REGEXP.test(this.pathname);
+  }
+
+  // TODO: Same method on the Relier flags, can we DRY?
+  isOAuth() {
+    return (
+      !!(
+        this._searchParam('client_id') ||
+        // verification
+        this._isOAuthVerificationSameBrowser()
+      ) ||
+      !!this._isOAuthVerificationDifferentBrowser() ||
+      // any URL with 'oauth' in the path.
+      /oauth/.test(this.pathname)
```

Comment:
this seems to check for "oauth" _anywhere_ in the pathname. Should this have some extra slashes or something like `/\/oauth\//` to make it a bit more restrictive?

Reply:
Our paths that are included here are `/oauth`, `/oauth/force_auth`, `/oauth/signin`, `/oauth/signup`, and `/oauth/success/:client_id`. Maybe a trailing slash in the front would be alright but not at the end? I'm kind of inclined to leave it, the most restrictive way is of course to create a whitelist with those paths mentioned, but ehh.

Reply:
Yeah, we could do `/^\/oauth(\/|$)/`, but that's kinda ugly: https://regexr.com/7c7p2
Or even `this.pathname.startsWith("/oauth")`.

But I'm guessing the risk of us adding "/oauthors/*" or similarly weird routes in the future is below 0%.

Reply:
Just an opinion, but when porting code, I think we should change things as little possible. It's already hard to spot errors after porting just due to the fact that these generally end up being large commits. Small changes can makes it even harder. I like the idea of improving this regex's but maybe it should be in a second pass. Again, totally an opinion, and not a very strong one.